legate/core/_lib/context.pyx

```diff
@@ -0,0 +1,88 @@
+# Copyright 2023 NVIDIA Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# distutils: language = c++
+
+import cython
+from libcpp cimport bool
+from libcpp.string cimport string
+
+
+cdef extern from "core/legate_c.h" nogil:
+    ctypedef enum legate_core_variant_t:
+        pass
+
+cdef extern from "core/task/task_info.h" namespace "legate" nogil:
+    cdef cppclass TaskInfo:
+        bool has_variant(int)
+
+cdef extern from "core/runtime/context.h" namespace "legate" nogil:
+    cdef cppclass LibraryContext:
+        unsigned int get_task_id(long long)
+        unsigned int get_mapper_id(long long)
+        int get_reduction_op_id(long long)
+        unsigned int get_projection_id(long long)
+        unsigned int get_sharding_id(long long)
+        TaskInfo* find_task(long long)
+
+cdef extern from "core/runtime/runtime.h" namespace "legate" nogil:
+    cdef cppclass Runtime:
+        @staticmethod
+        Runtime* get_runtime()
+        LibraryContext* find_library(string, bool)
+
+
+cdef class CppTaskInfo:
+    cdef const TaskInfo* _task_info
+
+    @staticmethod
+    cdef CppTaskInfo from_ptr(const TaskInfo* p_task_info):
+        cdef CppTaskInfo result = CppTaskInfo.__new__(CppTaskInfo)
+        result._task_info = p_task_info
+        return result
+
+    @property
+    def valid(self) -> bool:
+        return self._task_info != NULL
+
+    def has_variant(self, int variant_id) -> bool:
+        return self._task_info.has_variant(
+            cython.cast(legate_core_variant_t, variant_id)
+        )
+
+
+cdef class Context:
+    cdef LibraryContext* _context
+
+    def __init__(self, str library_name, bool can_fail=False) -> None:
```

Comment:
Better to use `__cinit__`, rather than `__init__` for C++-level initialization of the object:

https://cython.readthedocs.io/en/latest/src/userguide/special_methods.html#initialisation-methods-cinit-and-init